### PR TITLE
fix(gate): Improve logging and responses, prepare 0.4.1

### PR DESCRIPTION
**Pull request by Natoandro**
- Logging before and after each faillible operation
  -  Runtimes: foreign resolvers
- Always log before reporting error: HTTP response
- Fix error code in artifact_service
- Add `BaseError` class for structured messages in responses

## Summary by CodeRabbit

- **New Features**
  - Updated Docker image versions and dependency versions to ensure compatibility and stability.
  - Added a search functionality to the app.

- **Bug Fixes**
  - Enhanced error handling with specific error classes for more detailed error messages.

- **Refactor**
  - Replaced generic `Error` instances with specific error classes for better error categorization.
  - Refactored error handling in HTTP response functions to use a `BaseError` class.

- **Chores**
  - Updated version numbers across multiple configuration files to `0.4.1-0`.

### CODE REVIEW

**Review by Yohe-Am — DISMISSED** [review]

File: typegate/src/errors.ts [hunk 1]
```diff
@@ -1,6 +1,141 @@
...
+  toResponse(graphqlFormat = true): Response {
+    const type = this.#type ?? this.constructor.name;
+    logger.error(
+      "{}[{}:{}]: {}",
+      type,
+      this.kind,
+      this.module ?? "",
+      this.message,
+    );
+    logger.warn("Responding with HTTP {}", this.code);
```
Comment (afmika) [inline typegate/src/errors.ts:66]: 👍

**Review by afmika — APPROVED** [review]

File: meta-cli/Cargo.toml [hunk 2]
```diff
@@ -29,72 +29,98 @@ default = []
 typegate = ["dep:typegate_engine"]
 
 [dependencies]
+# internal
```
Comment (zifeo) [inline meta-cli/Cargo.toml:32]: 👏

File: typegate/src/errors.ts [hunk 3]
```diff
@@ -1,6 +1,141 @@
...
+  }
+
+  withType(type: string): this {
+    this.#type = type;
+    return this;
+  }
+
+  toResponse(graphqlFormat = true): Response {
+    const type = this.#type ?? this.constructor.name;
+    logger.error(
```
Comment (zifeo) [inline typegate/src/errors.ts:59]: ideally, in non debug (config) we shall hide the details of the error to the user but that might also reduce the ux (for now at least)

**Review by zifeo — DISMISSED** [review]
We will likely have to improve log even better, what a super start 💪

File: typegate/src/errors.ts (hunk 3 above)
Comment (Yohe-Am) [reply typegate/src/errors.ts:59]: Yeah, the current approach is very...unpolished like that. Something consistent and more crafted is in order.

**Review by Natoandro — COMMENTED** [review]
Great :+1: 
LGTM

**Pull request merged**